mountpoint-s3/src/fs.rs

```diff
@@ -603,16 +613,29 @@ where
         ino: InodeNo,
         atime: Option<OffsetDateTime>,
         mtime: Option<OffsetDateTime>,
+        _size: Option<u64>,
         _flags: Option<u32>,
     ) -> Result<Attr, Error> {
         tracing::info!(
-            "fs:setattr with ino {:?} flags {:?} atime {:?} mtime {:?}",
+            "fs:setattr with ino {:?} flags {:?} atime {:?} mtime {:?} size {:?}",
             ino,
             _flags,
             atime,
-            mtime
+            mtime,
+            _size
         );
-        let lookup = self.superblock.setattr(&self.client, ino, atime, mtime).await?;
+        let setattr_result = self.superblock.setattr(&self.client, ino, atime, mtime).await;
+        let lookup = match (setattr_result, _size) {
+            (Ok(lookup), _) => lookup,
+            (Err(InodeError::SetAttrNotPermittedOnRemoteInode(_)), Some(_)) => {
+                // For file overwrite(in case of not having `--file-overwrite` flag), first `setattr()` is called with `Some` value of size, whereas for only metadata modification size is `None`.
+                // In both cases, `setattr()` fails if the call is made on remote file.
+                // So, distinguishing file overwrite error from general metadata modification for remote file error.
+                return Err(
+                    err!(libc::EPERM, "file overwrite is disabled by default, you need to remount with --allow-overwrite flag and open the file in truncate mode (O_TRUNC) to overwrite it"));
+            }
```

Comment:
What I want to call out here is that we are explicitly making the trade-off: if we see that size is set, then we assume that it is an attempt to make a truncate overwrite.

Maybe something like...

```rust
// We want to provide better feedback to users to prompt them to opt-in to file overwrites if it looks like what the application needs.
// Instead of complex logic to match `setattr` truncation only, we just check for the error and if the size was set in the request.
// If so, we assume its probably a truncation.
```

Two questions on this logic though:

- Does the value of the size matter? Will size ever be non-zero for overwrites that we support? We could explicitly match on the value too if it makes sense.
- Why do we not check if overwrites is actually enabled or not? This error doesn't make sense if customer has it enabled.

Reply:
Unfortunately, I could find any good documentation of `setattr()` method of fuser and what value is set for size for different cases. Through my observations, it is always `Some(0)` for file overwrite and `None` for other cases. 
We can add the check if the overwrite is enabled or not. I did not add it, because when enabled, it does not call setattr. Fuse follows lookup -> open -> getxattr (not supported) -> write -> flush.

Reply:
There is a subtle detail here. If overwrite is enabled mountpoint will be initialized with `FUSE_ATOMIC_O_TRUNC` capability, that means the kernel will not call setattr on truncate. But at least we should add a comment why we don't need to check if overwrite is enabled here. https://github.com/awslabs/mountpoint-s3/blob/e52d68ede985ecdb31f68e8db2beef29b528c8b1/mountpoint-s3/src/fs.rs#L523-L532

Reply:
Thanks @monthonk @dannycjones . I have added the check for overwrite in `setattr` as well just to be on the safe side even when there is any implementation change in future.

Reply:
Makes sense, thanks Ankit! That sounds right.